exporter/loadbalancingexporter/resolver_srv.go

```diff
@@ -0,0 +1,252 @@
+		return nil, errNoSRV
+	}
+	if interval == 0 {
+		interval = defaultResInterval
+	}
+	if timeout == 0 {
+		timeout = defaultResTimeout
+	}
+
+	service, proto, name, err := parseSRVHostname(srvHostname)
+	if err != nil {
+		logger.Warn("failed to parse SRV hostname", zap.Error(err))
+	}
+
+	return &srvResolver{
+		logger:      logger,
+		srvService:  service,
+		srvProto:    proto,
+		srvName:     name,
+		port:        port,
+		resolver:    &net.Resolver{},
+		resInterval: interval,
+		resTimeout:  timeout,
+		stopCh:      make(chan struct{}),
+	}, nil
+}
+
+func (r *srvResolver) start(ctx context.Context) error {
+	if _, err := r.resolve(ctx); err != nil {
+		r.logger.Warn("failed to resolve", zap.Error(err))
+	}
+
+	go r.periodicallyResolve()
+
+	r.logger.Debug("SRV resolver started",
+		zap.String("SRV name", r.srvName), zap.String("port", r.port),
+		zap.Duration("interval", r.resInterval), zap.Duration("timeout", r.resTimeout))
+	return nil
+}
+
+func (r *srvResolver) shutdown(_ context.Context) error {
+	r.changeCallbackLock.Lock()
+	r.onChangeCallbacks = nil
+	r.changeCallbackLock.Unlock()
+
+	close(r.stopCh)
+	r.shutdownWg.Wait()
+	return nil
+}
+
+func (r *srvResolver) periodicallyResolve() {
+	ticker := time.NewTicker(r.resInterval)
+
+	for {
+		select {
+		case <-ticker.C:
+			ctx, cancel := context.WithTimeout(context.Background(), r.resTimeout)
+			if _, err := r.resolve(ctx); err != nil {
+				r.logger.Warn("failed to resolve", zap.Error(err))
+			} else {
+				r.logger.Debug("resolved successfully")
+			}
+			cancel()
+		case <-r.stopCh:
+			return
+		}
+	}
+}
+
+func (r *srvResolver) resolve(ctx context.Context) ([]string, error) {
+	r.shutdownWg.Add(1)
+	defer r.shutdownWg.Done()
+
+	_, srvs, err := r.resolver.LookupSRV(ctx, r.srvService, r.srvProto, r.srvName)
+	if err != nil {
+		_ = stats.RecordWithTags(ctx, srvResolverSuccessFalseMutators, mNumResolutions.M(1))
+		return nil, err
+	}
+
+	_ = stats.RecordWithTags(ctx, srvResolverSuccessTrueMutators, mNumResolutions.M(1))
+
+	// backendsWithIPs tracks the IP addresses for changes
+	backendsWithIPs := make(map[string]string)
+	for _, srv := range srvs {
+		target := strings.TrimSuffix(srv.Target, ".")
+		backendsWithIPs[target] = ""
+	}
+	// backends is what we use to compare against the current endpoints
+	var backends []string
+
+	// Lookup the IP addresses for the A records
+	for aRec := range backendsWithIPs {
+
+		// handle backends first
+		backend := aRec
+		// if a port is specified in the configuration, add it
+		if r.port != "" {
+			backend = fmt.Sprintf("%s:%s", backend, r.port)
+		}
+		backends = append(backends, backend)
+
+		ips, err := r.resolver.LookupIPAddr(ctx, aRec)
+		// Return the A record. If we can't resolve them, we'll try again next iteration
+		if err != nil {
+			_ = stats.RecordWithTags(ctx, srvResolverSuccessFalseMutators, mNumResolutions.M(1))
+			continue
+		}
+		// A headless Service SRV target only returns 1 IP address for its A record
```

Comment:
This might be true for Kubernetes, but this resolver isn't only for Kubernetes, right?

Reply:
Double-checking the documentation, this is what I see:

>SRV Records are created for named ports that are part of normal or headless services. For each named port, the SRV record has the form _port-name._port-protocol.my-svc.my-namespace.svc.cluster-domain.example. For a regular Service, this resolves to the port number and the domain name: my-svc.my-namespace.svc.cluster-domain.example. For a headless Service, this resolves to multiple answers, one for each Pod that is backing the Service, and contains the port number and the domain name of the Pod of the form hostname.my-svc.my-namespace.svc.cluster-domain.example.

https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#srv-records

I believe this resolver should target generic SRV records, not only the ones that Kubernetes generates for headless services.

Reply:
I think this is another area where I should note down a limitation until priority and weight are introduced. Is my thinking correct here?
- For a k8s headless service there is a direct 1:1 mapping
- For a normal service it is going to return a single IP but that's the service VIP and could get load balanced to any n number of endpoints and if n > 1 you might have spans delivered to different endpoints
- Outside of k8s I'd say the guidance would be to use the weight feature when it is in place. If you have a record which returns back multiple IPs then use the normal resolver. Otherwise make sure your targets only return back A records with single IPs and multiple results are handled inside of the SRV record.

If we did want this implemented then perhaps we should go back to the original name mentioned in the issue and make an issue to create `DNSSRV` since that's what it does for those other apps. We would essentially say if your desire is to get back a hostname as opposed to IP(s) then the underlying targets must have a 1:1 mapping.

Reply:
If we are mimicking the dnssrvnoa behavior from Thanos/Loki/Mimir/Cortex, that's the name we should use for the resolver, IMO.